x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/HashJoinExec.java

```diff
@@ -91,21 +93,27 @@ public List<Attribute> rightFields() {
 
     public Set<Attribute> addedFields() {
         if (lazyAddedFields == null) {
-            lazyAddedFields = new AttributeSet(output());
-            lazyAddedFields.removeAll(left().output());
+            lazyAddedFields = new AttributeSet(addedFields);
         }
         return lazyAddedFields;
     }
 
     @Override
     public PhysicalPlan estimateRowSize(State state) {
-        state.add(false, output);
+        state.add(false, addedFields);
         return this;
     }
 
     @Override
     public List<Attribute> output() {
-        return output;
+        if (lazyOutput == null) {
+            lazyOutput = new ArrayList<>(left().output());
+            var rightFieldNames = rightFields.stream().map(Attribute::name).toList();
+            lazyOutput.removeIf(a -> rightFieldNames.contains(a.name()));
+            lazyOutput.addAll(addedFields);
+            lazyOutput.addAll(rightFields);
```

Comment:
nit: this looks like we're preserving the join keys from the right, but we normally preserve join keys from the left. In this case it doesn't matter, because they are the same, but this differs conceptually from LOOKUP JOIN which explicitly preserves the join keys from the left.